File: compare_sklearn_glmnet.py
from scipy.io import mmwrite, mmread
from sklearn.linear_model.coordinate_descent import _alpha_grid
from sklearn.linear_model import lasso_path
from 
from time import time

# Newsgroup datasets
X_new = mmread("X_new.mtx")
y_new = mmread("y_new.mtx")
y_new = y_new.toarray()[0]
#alphas = _alpha_grid(X_new, y_new, eps=1e-3, fit_intercept=True, normalize=False, n_alphas=100)
t = time()
lasso_path(X_new, y_new, eps=1e-3, precompute=False, fit_intercept=True, normalize=False, n_alphas=100)
print time() - t

X_new = mmread("haxby_X.mtx").toarray()
y_new = mmread("haxby_y.mtx").toarray()[0]
#print y_new.shape
#y_new = y_new.toarray()[0]
# alphas = _alpha_grid(X_new, y_new, eps=1e-3, fit_intercept=True, normalize=False, n_alphas=100)

t = time()
coef = lasso_path(X_new, y_new, eps=1e-3, precompute=False, fit_intercept=True, normalize=False, alphas=alphas)#n_alphas=100)
coef_ = coef[1]
sq_loss = np.sum(0.5*(y_new[:, np.newaxis] - safe_sparse_dot(X_new, coef_))**2, axis=0)
l2_penalty = np.sum(coef_**2, axis=0)
l1_penalty = np.sum(np.abs(coef_), axis=0)

print time() - t

File: compare_sklearn_glmnet.R
library(glmnet)
library(Matrix)


# X_new <- readMM("X_new.mtx")
# y_new <- Matrix(readMM("y_new.mtx"))
# ptm <- proc.time()
# fit <- glmnet(X_new, y_new, lambda.min.ratio=0.001, standardize=FALSE, intercept=TRUE)
# proc.time() - ptm
# print(fit)

X_new <- Matrix(readMM("haxby_X.mtx"))
y_new <- Matrix(readMM("haxby_y.mtx"))[1,]
ptm <- proc.time()
glmnet.control(fdev = 0)
fit <- glmnet(X_new, y_new, lambda.min.ratio=0.001, standardize=FALSE,
              intercept=TRUE, maxit=10**7, thresh=1e-7)
fit
proc.time() - ptm

coef_ <- fit$beta
sqloss <- 0.5 * colSums((y_new - (X_new %*% coef_))**2)
l1_penalty <- colSums(abs(c))
l2_penalty <- colSums(c**2) * 0.5
sqloss + l1_penalty + l2_penalty
# t_ = coef(fit, s=0.007202, exact=TRUE)
# which(t_!=0)
# t_[t_ != 0]

# array([[  6.92123804e-17,   2.83210732e-02],
#        [  0.00000000e+00,   5.75088994e-03],
#        [  0.00000000e+00,   2.54628035e-03]])
